File: ice0208/dp-1d/1137.py
class Solution:
    def __init__(self):
        self.cach = dict()

    def tribonacci(self, n: int) -> int:
        if n == 0:
            return 0
        if n == 1 or n == 2:
            return 1

        if n not in self.cach:
            self.cach[n] = (
                self.tribonacci(n - 1) + self.tribonacci(n - 2) + self.tribonacci(n - 3)
            )

        return self.cach[n]


if __name__ == "__main__":
    print(Solution().tribonacci(4))

File: moonpoison/dp_1demensional/1137.cpp
#include<bits/stdc++.h>
using namespace std;
class Solution {
public:
    int dp[38]={0};
    int func(int n)
    {
        if(n==0) return 0;
        else if(dp[n]!=0) return dp[n];
        else 
        {
            dp[n]=func(n-3)+func(n-2)+func(n-1);
            return dp[n];
        }
    }
    int tribonacci(int n) 
    {
        dp[0]=0; dp[1]=1; dp[2]=1;
        return func(n);
    }
};